lib/chef/knife/bootstrap/train_connector.rb

```diff
@@ -0,0 +1,287 @@
+# Copyright:: Copyright (c) 2019 Chef Software Inc.
+#
+# License:: Apache License, Version 2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+require "train"
+require "tempfile"
+require "uri"
+
+class Chef
+  class Knife
+    class Bootstrap < Knife
+      class TrainConnector
+        SSH_CONFIG_OVERRIDE_KEYS = [:user, :port, :proxy].freeze
+
+        MKTEMP_WIN_COMMAND = <<~EOM.freeze
+          $parent = [System.IO.Path]::GetTempPath();
+          [string] $name = [System.Guid]::NewGuid();
+          $tmp = New-Item -ItemType Directory -Path;
+          (Join-Path $parent $name);
+          $tmp.FullName
+        EOM
+
+        MKTEMP_NIX_COMMAND = "bash -c 'd=$(mktemp -d ${TMPDIR:-/tmp}/chef_XXXXXX); echo $d'".freeze
+
+        def initialize(host_url, default_transport, opts)
+          uri_opts = opts_from_uri(host_url)
+          uri_opts[:backend] ||= @default_transport
+          @transport_type = uri_opts[:backend]
+
+          # opts in the URI will override user-provided options
+          @config = transport_config(host_url, opts.merge(uri_opts))
+
+          Train.validate_backend(@config)
+          @train = Train.create(@transport_type, @config)
```

Comment:
you might want to consider making this lazy accessors for `@train ||= ...` and `@config ||= ...` just  on the basis of removing code from the initializer